src/Microsoft.ML.Recommender/MatrixFactorizationStatic.cs

```diff
@@ -0,0 +1,125 @@
+                    args =>
+                    {
+                        advancedSettings?.Invoke(args);
+                        args.Lambda = regularizationCoefficient;
+                        args.K = approximationRank;
+                        args.Eta = learningRate;
+                        args.NumIterations = numIterations;
+                    });
+                if (onFit != null)
+                    return trainer.WithOnFitDelegate(trans => onFit(trans.Model));
+                else
+                    return trainer;
+            }, label, matrixColumnIndex, matrixRowIndex);
+            return rec.Output;
+        }
+
+        private sealed class MatrixFactorizationReconciler<T> : TrainerEstimatorReconciler
+        {
+            // Output column name of the trained estimator.
+            private static string FixedOutputName => DefaultColumnNames.Score;
+
+            // A function used to create trainer of matrix factorization. It instantiates a trainer by indicating the
+            // expected inputs and output (IDataView's) column names. That trainer has a Fit(IDataView data) for learning
+            // a MatrixFactorizationPredictionTransformer from the data.
+            private readonly Func<IHostEnvironment, string, string, string, IEstimator<ITransformer>> _factory;
+
+            /// <summary>
+            /// The only output produced by matrix factorization predictor
+            /// </summary>
+            public Scalar<float> Output { get; }
+
+            /// <summary>
+            /// The output columns.
+            /// </summary>
+            protected override IEnumerable<PipelineColumn> Outputs { get; }
+
+            public MatrixFactorizationReconciler(Func<IHostEnvironment, string, string, string, IEstimator<ITransformer>> factory,
+                Scalar<float> label, Key<T> matColumnIndex, Key<T> matRowIndex)
+                : base(MakeInputs(Contracts.CheckRef(label, nameof(label)), Contracts.CheckRef(matColumnIndex, nameof(matColumnIndex)), Contracts.CheckRef(matRowIndex, nameof(matRowIndex))),
+                      new string[] { FixedOutputName })
+            {
+                Contracts.AssertValue(factory);
+                _factory = factory;
+
+                Output = new Impl(this);
+                Outputs = new PipelineColumn[] { Output };
+            }
+
+            private static PipelineColumn[] MakeInputs(Scalar<float> label, PipelineColumn matrixRowIndex, PipelineColumn matrixColumnIndex)
+                => new PipelineColumn[] { label, matrixRowIndex, matrixColumnIndex };
+
+            protected override IEstimator<ITransformer> ReconcileCore(IHostEnvironment env, string[] inputNames)
+            {
+                Contracts.AssertValue(env);
+
+                // The first, second, third names are label, matrix's row index, and matrix's column index, respectively.
```

Comment:
>are label, matrix's row index, and matrix's column index [](start = 50, length = 56)

>Contracts.CheckRef(label, nameof(label)), Contracts.CheckRef(matColumnIndex, nameof(matColumnIndex)), Contracts.CheckRef(matRowIndex, nameof(matRowIndex))

It's label, column index, row index.
At least in your code.

Reply:
Good catch. Thanks.

---
In reply to: [229011380](https://github.com/dotnet/machinelearning/pull/1407#discussion_r229011380) [](ancestors = 229011380)